File: producer/src/events.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Copy, Serialize, Deserialize)]
pub enum Country {
    #[serde(rename = "US")]
    US,

    #[serde(rename = "NG")]
    NG,

    #[serde(rename = "GB")]
    GB,

    #[serde(rename = "CA")]
    CA,

    #[serde(rename = "FR")]
    FR,

    #[serde(rename = "ES")]
    ES,
}

#[derive(Clone, Debug, Copy, Serialize, Deserialize)]
pub enum Currency {
    #[serde(rename = "USD")]
    USD,

    #[serde(rename = "NGN")]
    NGN,

    #[serde(rename = "EUR")]
    EUR,

    #[serde(rename = "GBP")]
    GBP,

    #[serde(rename = "CAD")]
    CAD,
}

#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
pub enum TransactionEventType {
    #[serde(rename = "transaction.initiated")]
    Initiated,

    #[serde(rename = "transaction.authorized")]
    Authorized,

    #[serde(rename = "transaction.completed")]
    Completed,

    #[serde(rename = "transaction.declined")]
    Declined,

    #[serde(rename = "transaction.flagged")]
    Flagged,

    #[serde(rename = "transaction.reversed")]
    Reversed,

    #[serde(rename = "transaction.expired")]
    Expired,
}

#[derive(Clone)]
pub struct Money {
    pub amount: u64,
    pub currency: Currency,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct TransactionEvent {
    pub event_id: String,
    pub idempotency_key: String,
    pub event_time: String,
    pub event_type: TransactionEventType,

    pub customer_id: String,
    pub amount: u64,
    pub currency: Currency,
    pub merchant_id: String,
    pub country: Country,

    pub instrument_fingerprint: String,
}

File: processor/internal/transactions/events.go
package transactions

type TransactionEventType string

const (
	TransactionInitiated  TransactionEventType = "transaction.initiated"
	TransactionAuthorized TransactionEventType = "transaction.authorized"
	TransactionCompleted  TransactionEventType = "transaction.completed"
	TransactionDeclined   TransactionEventType = "transaction.declined"
	TransactionFlagged    TransactionEventType = "transaction.flagged"
	TransactionReversed   TransactionEventType = "transaction.reversed"
	TransactionExpired    TransactionEventType = "transaction.expired"
)

type Currency string

const (
	USD Currency = "USD"
	NGN Currency = "NGN"
	EUR Currency = "EUR"
	GBP Currency = "GBP"
	CAD Currency = "CAD"
)

type ValidationStatus string

const (
	ValidationStatusApproved ValidationStatus = "approved"
	ValidationStatusFlagged  ValidationStatus = "flagged"
	ValidationStatusRejected ValidationStatus = "rejected"
)

type RawTransactionEvent struct {
	EventID        string               `json:"event_id"`
	IdempotencyKey string               `json:"idempotency_key"`
	EventTime      string               `json:"event_time"`
	EventType      TransactionEventType `json:"event_type"`

	CustomerID            string   `json:"customer_id"`
	Amount                uint64   `json:"amount"`
	Currency              Currency `json:"currency"`
	MerchantID            string   `json:"merchant_id"`
	Country               string   `json:"country"`
	InstrumentFingerprint string   `json:"instrument_fingerprint"`
}

type ProcessedTransactionEvent struct {
	EventID        string               `json:"event_id"`
	IdempotencyKey string               `json:"idempotency_key"`
	EventTime      string               `json:"event_time"`
	EventType      TransactionEventType `json:"event_type"`

	CustomerID            string   `json:"customer_id"`
	Amount                uint64   `json:"amount"`
	Currency              Currency `json:"currency"`
	MerchantID            string   `json:"merchant_id"`
	Country               string   `json:"country"`
	InstrumentFingerprint string   `json:"instrument_fingerprint"`
	IsFraud               bool     `json:"is_fraud"`

	RiskScore           float64          `json:"risk_score"`
	ProcessedAt         string           `json:"processed_at"`
	ProcessorID         string           `json:"processor_id"`
	ProcessingLatencyMs int64            `json:"processing_latency_ms"`
	RiskFactors         []string         `json:"risk_factors"`
	ValidationStatus    ValidationStatus `json:"validation_status"`
}
